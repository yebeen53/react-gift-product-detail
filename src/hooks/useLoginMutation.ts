import { useMutation } from '@tanstack/react-query';
import { apiClient } from '@/api/apiClient';

interface LoginPayload {
  email: string;
  password: string;
}

interface LoginResponse {
  authToken: string;
  id: number;
  name: string;
  email: string;
}

export const useLoginMutation = () => {
  return useMutation({
    mutationFn: async (payload: LoginPayload): Promise<LoginResponse> => {
      const res = await apiClient.post('/api/login', payload);
      return res.data.data;
    },
  });
};