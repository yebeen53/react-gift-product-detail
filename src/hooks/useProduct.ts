import {
  useSuspenseQuery,
  useMutation,
  useQueryClient,
} from '@tanstack/react-query';
import {
  fetchProduct,
  fetchProductDetail,
  fetchHighlightReview,
  fetchWishCount,
} from '@/api/products/product';

export const useProduct = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['product', productId],
    queryFn: () => fetchProduct(productId),
  });

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });

export const useHighlightReview = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['highlightReview', productId],
    queryFn: () => fetchHighlightReview(productId),
  });

export const useWishCount = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['wishCount', productId],
    queryFn: () => fetchWishCount(productId),
  });

export const useWishMutation = (productId: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async () => Promise.resolve(), // 실제 API 요청 함수 넣기

    onMutate: async () => {
      await queryClient.cancelQueries({ queryKey: ['wishCount', productId] });

      const prevWish = queryClient.getQueryData<{
        wishCount: number;
        isWished: boolean;
      }>(['wishCount', productId]);

      if (prevWish) {
        queryClient.setQueryData(['wishCount', productId], {
          ...prevWish,

          wishCount: prevWish.isWished
            ? prevWish.wishCount - 1
            : prevWish.wishCount + 1,
          isWished: !prevWish.isWished, // 토글
        });
      }

      return { prevWish };
    },

    onError: (_err, _vars, context) => {
      if (context?.prevWish) {
        queryClient.setQueryData(['wishCount', productId], context.prevWish);
      }
    },

    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['wishCount', productId] });
    },
  });
};
