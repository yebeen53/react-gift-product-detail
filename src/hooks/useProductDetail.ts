import { useSuspenseQuery } from '@tanstack/react-query';
import axios from 'axios';

interface Announcement {
  name: string;
  value: string;
  displayOrder: number;
}

interface ProductDetail {
  description: string;
  announcements: Announcement[];
}

const fetchProductDetail = async (productId: string): Promise<ProductDetail> => {
  const response = await axios.get(`/api/products/${productId}/detail`);
  return response.data.data;
};

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });
