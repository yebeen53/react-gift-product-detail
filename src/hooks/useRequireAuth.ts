import useAuth from '@/hooks/useAuth';
import { useLocation, useNavigate } from 'react-router-dom';
import { useEffect } from 'react';
import { ROUTES } from '@/constants/routes';

const useRequireAuth = () => {
  const { user } = useAuth();
  const location = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    if (!user || !user.authToken) {
      navigate(ROUTES.HOME, {
        replace: true,
        state: { from: location.pathname },
      });
    }
  }, [user, navigate, location]);

  return user;
};

export default useRequireAuth;
