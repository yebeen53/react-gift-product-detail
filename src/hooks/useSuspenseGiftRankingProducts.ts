import { fetchRanking } from '@/api/products/ranking';
import { useSuspenseQuery } from '@tanstack/react-query';

export const useSuspenseGiftRankingProducts = (
  genderCode: string,
  categoryCode: string
) => {
  return useSuspenseQuery({
    queryKey: ['giftRankingProducts', genderCode, categoryCode],
    queryFn: () => fetchRanking(genderCode, categoryCode),
  });
};
