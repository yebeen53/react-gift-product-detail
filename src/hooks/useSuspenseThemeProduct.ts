import {
  useSuspenseInfiniteQuery,
  type QueryFunctionContext,
} from '@tanstack/react-query';
import axios from 'axios';
import type { ThemeProductsData } from '@/types/product';

const fetchThemeProducts = async (
  ctx: QueryFunctionContext<readonly [string, string], number>
): Promise<ThemeProductsData> => {
  const [, themeId] = ctx.queryKey;
  const pageParam = ctx.pageParam ?? 0;

  const res = await axios.get(`/api/themes/${themeId}/products`, {
    params: { cursor: pageParam, limit: 10 },
  });
  const data = res.data.data;

  return {
    list: data.list,
    cursor: data.cursor,
    hasMoreList: data.hasMoreList,
  };
};

export const useSuspenseThemeProducts = (themeId: string) => {
  return useSuspenseInfiniteQuery({
    queryKey: ['themeProducts', themeId] as const,
    queryFn: fetchThemeProducts,
    initialPageParam: 0,
    getNextPageParam: (lastPage) =>
      lastPage.hasMoreList ? lastPage.cursor : undefined,
  });
};
