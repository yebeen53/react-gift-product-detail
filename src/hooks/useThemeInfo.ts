import { apiClient } from '@/api/apiClient';
import { useSuspenseQuery } from '@tanstack/react-query';

export interface ThemeDetail {
  image: string | undefined;
  name: string;
  themeId: string;
  title: string;
  description: string | null;
  backgroundColor: string;
}

const fetchThemeDetail = async (themeId: string): Promise<ThemeDetail> => {
  const res = await apiClient.get(`/api/themes/${themeId}/info`);
  const data = res.data?.data;
  if (!data) {
    throw new Error('테마 정보가 없습니다.');
  }
  return data;
};

export const useThemeInfo = (themeId: string) => {
  return useSuspenseQuery({
    queryKey: ['themeDetail', themeId],
    queryFn: () => fetchThemeDetail(themeId),
  });
};
