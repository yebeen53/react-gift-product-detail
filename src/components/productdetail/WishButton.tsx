import { useWishCount, useWishMutation } from '@/hooks/useProduct';
import { Heart } from 'lucide-react';
import { css } from '@emotion/react';
import type { Theme } from '@/data/theme';
import theme from '@/data/theme';

interface Props {
  productId: string;
}

const WishButton = ({ productId }: Props) => {
  const { data } = useWishCount(productId);
  const mutation = useWishMutation(productId);

  const wishCount = data?.wishCount ?? 0;
  const isWished = data?.isWished ?? false;

  const handleClick = () => {
    if (!mutation.isPending) {
      mutation.mutate();
    }
  };

  return (
    <button onClick={handleClick} css={buttonStyle(theme)} aria-pressed={isWished}>
      <Heart size={25} fill={isWished ? 'red' : 'none'} color="red" />
      <span>{wishCount}</span>
    </button>
  );
};

const buttonStyle =(theme:Theme)=> css`
  display: flex;
  align-items: center;
  gap: ${theme.spacing.spacing4};
  background: transparent;
  border: none;
  cursor: pointer;
`;

export default WishButton;
