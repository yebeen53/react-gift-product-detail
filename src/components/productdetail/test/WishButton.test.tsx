import { render, screen, fireEvent } from '@testing-library/react';
import WishButton from '../WishButton'; 
import { useWishCount, useWishMutation } from '@/hooks/useProduct';

jest.mock('@/hooks/useProduct', () => ({
  useWishCount: jest.fn(),
  useWishMutation: jest.fn(),
}));

describe('WishButton 컴포넌트', () => {
  const mockMutate = jest.fn();

  beforeEach(() => {
    
    jest.clearAllMocks();
    (useWishCount as jest.Mock).mockReturnValue({
      data: { wishCount: 10, isWished: false },
      isLoading: false,
      isSuccess: true,
      isError: false,
      error: null,
    });

    (useWishMutation as jest.Mock).mockReturnValue({
      mutate: mockMutate,
      isPending: false,
      isSuccess: false,
      isError: false,
      data: undefined,
      error: null,
      reset: jest.fn(),
      status: 'idle',
      variables: undefined,
    });
  });

  it('찜하기 버튼과 찜 카운트가 올바르게 렌더링되어야 한다 (찜 안 한 상태)', () => {
    render(<WishButton productId="p1" />); 

    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(screen.getByText('10')).toBeInTheDocument();
    expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'false');
    expect(screen.getByRole('button').querySelector('svg')).toHaveAttribute(
      'fill',
      'none'
    );
  });

  it('찜하기 버튼과 찜 카운트가 올바르게 렌더링되어야 한다 (찜 한 상태)', () => {
    (useWishCount as jest.Mock).mockReturnValue({
      data: { wishCount: 15, isWished: true },
    });
    render(<WishButton productId="p1" />);

    expect(screen.getByText('15')).toBeInTheDocument();
    expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'true');
    expect(screen.getByRole('button').querySelector('svg')).toHaveAttribute(
      'fill',
      'red'
    );
  });

  it('찜하기 버튼 클릭 시 mutate 함수가 호출되어야 한다', () => {
    render(<WishButton productId="p1" />); 

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(mockMutate).toHaveBeenCalledTimes(1);
  });

  it('mutation이 진행 중일 때는 버튼 클릭 시 mutate 함수가 호출되지 않아야 한다', () => {
    (useWishMutation as jest.Mock).mockReturnValue({
      mutate: mockMutate,
      isPending: true, 
      isSuccess: false,
      isError: false,
      data: undefined,
      error: null,
      reset: jest.fn(),
      status: 'pending', 
      variables: undefined,
    });
    render(<WishButton productId="p1" />); 

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(mockMutate).not.toHaveBeenCalled();
  });

  it('컴포넌트의 스타일이 스냅샷과 일치해야 한다', () => {
    (useWishCount as jest.Mock).mockReturnValue({
      data: { wishCount: 0, isWished: false },
    });
    const { asFragment } = render(<WishButton productId="p1" />); 
    expect(asFragment()).toMatchSnapshot();
  });

  it('찜한 상태일 때 컴포넌트의 스타일이 스냅샷과 일치해야 한다', () => {
    (useWishCount as jest.Mock).mockReturnValue({
      data: { wishCount: 1, isWished: true },
    });
    const { asFragment } = render(<WishButton productId="p1" />); 
    expect(asFragment()).toMatchSnapshot();
  });
});

