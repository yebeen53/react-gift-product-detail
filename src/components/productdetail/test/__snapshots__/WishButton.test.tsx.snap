// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`WishButton 컴포넌트 찜한 상태일 때 컴포넌트의 스타일이 스냅샷과 일치해야 한다 1`] = `
<DocumentFragment>
  <button
    aria-pressed="true"
    css="[object Object]"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-heart"
      fill="red"
      height="25"
      stroke="red"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="25"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
      />
    </svg>
    <span>
      1
    </span>
  </button>
</DocumentFragment>
`;

exports[`WishButton 컴포넌트 컴포넌트의 스타일이 스냅샷과 일치해야 한다 1`] = `
<DocumentFragment>
  <button
    aria-pressed="false"
    css="[object Object]"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-heart"
      fill="none"
      height="25"
      stroke="red"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="25"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
      />
    </svg>
    <span>
      0
    </span>
  </button>
</DocumentFragment>
`;
